core/chains/evm/logpoller/log_poller_internal_test.go

```diff
@@ -380,14 +388,19 @@ func TestLogPoller_Replay(t *testing.T) {
 		}()
 		select {
 		case <-timeout:
-			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 1s")
+			assert.Fail(t, "lp.run() got stuck--failed to respond to second replay event within 2s")
 		case <-utils.WaitGroupChan(&wg):
 			lp.cancel()
 		}
 		<-done
 	})
 
-	lp.wg.Wait() // ensure logpoller has exited before continuing on to next test
+	timeout = time.After(4 * time.Second)
+	select {
+	case <-utils.WaitGroupChan(&lp.wg):
+	case <-timeout:
+		require.Fail(t, "Log poller failed to shutdown gracefully")
```

Comment:
It is incorrect by definition to return before these goroutines have exited. If they do any logging, it can cause additional errors and data races.

Reply:
Hmmm... so what's the proper way to detect a leak goroutine?  Previously we were just letting the test run for 10 mins until it times out... I figured this would be nicer, so we don't have to wait the full 10 mins.  Any suggestions?

Reply:
This is the goroutine that was leaking:
```
func (lp *logPoller) recvReplayComplete() {
    err := <-lp.replayComplete
    if err != nil {
        lp.lggr.Error(err)
    }
    lp.wg.Done()
}
```
Now that the replayComplete is sent, it should be able to exit.  But if for some reason it does get stuck (or another goroutine does?) is there a way to just forcibly kill it?

Reply:
Run it for less than 10 minutes?

Reply:
> is there a way to just forcibly kill it?

Can it just receive from the stopCh/ctx.Done() chan too?

Reply:
> Run it for less than 10 minutes?

I'm not sure what you mean by that.

But come to think of it, even if there were a way to do that (isn't that standard across all tests, is there a way to control it?) this would still be much better because it gives the user an error message stating exactly what the problem is.

The real problem is that when you see it flake in CI due to a deadlock, there's no indication at all of what the problem was... the only error message is "test timed out after 10 mins".   The only way to know which subtest was involved and where it got stuck is to drop into a debugger and stop the threads while it's running and inspect all the different threads to see what's waiting on what.

Timing out and printing an error message is much better behavior.  Even if there are other messages or warnings that get printed afterwards (eg about a goroutine printing a message after the test exits), it should still be pretty clear what the root cause was, and where it happened... thanks to this timeout.

Reply:
> Can it just receive from the stopCh/ctx.Done() chan too?

It wouldn't make much sense in production, since `lp.Close()` sends both signals at once (lp.replayComplete to let this thread know it can exit, and cancelling lp.ctx to let the runloop (which is the one that "owns" that context) know it can exit.  Also, the first thing the runloop does when it notices the context is cancelled is do a non-blocking send on lp.replayComplete... so the shutdown signal gets relayed in two different ways already.  Adding a third is unnecessary redundancy and goes against the go paradigm of having each thread only worry about its own context.

The only reason lp.replayComplete had to be sent explicitly by the test is because it doesn't call lp.Close() and doesn't start up lp.run()... starting lp.run() introduces too much uncertainty because of all the random tickers involved (each with their own jitter).

Reply:
> I'm not sure what you mean by that.

You are fully in control of the `-timeout` already. Adding another internal knob that is disconnected from this does not add any value IMHO.

> this would still be much better because it gives the user an error message stating exactly what the problem is.

Not if it also induces additional errors and can corrupt or crash the program from a data race.

> The more important issue this addresses is that when you see it flake in CI due to a deadlock, there's no indication at all of what the problem was... the only error message is "test timed out after 10 mins". The only way to know which subtest was involved and where it got stuck is to drop into a debugger and stop the threads while it's running and inspect all the different threads to see what's waiting on what.

Doesn't the timeout include stack traces of running goroutines?

> Timing out and printing an error message is much better behavior. Even if there are other messages or warnings that get printed afterwards (eg about a goroutine printing a message after the test exits), it should still be pretty clear what the root cause was, and where it happened... thanks to this timeout.

These are not warnings to be dismissed. They corrupt the test output and have undefined behavior.

Reply:
> You are fully in control of the -timeout already

I see, you were talking about running it locally...  I was thinking mostly about identifying failures in CI.

> Not if it also induces additional errors and can corrupt or crash the program from a data race.

I think I see your point.  I agree that if it can interfere with other running tests, then exiting in response to a stuck goroutine is worse than than just riding it out.

If it was the same goroutine stuck as last time, then I doubt there's any way it could log a message before the entire test times out anyway.  But we don't know that it would be the same one.

At any rate, this code should be unreachable anyway... I'll remove it.